qdrant_docs_mcp/tools/importer.py

```diff
@@ -0,0 +1,380 @@
+import tempfile
+from collections.abc import Generator
+from dataclasses import dataclass
+from itertools import chain
+from pathlib import Path
+
+import requests
+import semver
+from mcp_server_qdrant.embeddings.fastembed import FastEmbedProvider
+from qdrant_client import QdrantClient, models
+from rich.progress import track
+
+from qdrant_docs_mcp import LIBRARIES_ROOT
+from qdrant_docs_mcp.tools.extractor import extract
+from qdrant_docs_mcp.tools.models import (
+    Library,
+    LibraryConfig,
+    PartialSnippet,
+    Snippet,
+    SourceConfig,
+    SourceType,
+    VersionType,
+    get_default_config,
+)
+
+
+def _get_library_by_name(name: str) -> Library:
+    config_file = LIBRARIES_ROOT / f"{name}.json"
+
+    if not config_file.is_file():
+        raise FileNotFoundError(f'No configuration found for library "{name}"')
+
+    data = config_file.read_text()
+
+    return Library.model_validate_json(data)
+
+
+def _get_latest_github_tag(url: str) -> str | None:
+    user, repo = url.split("/")[-2:]
+    r = requests.get(f"https://api.github.com/repos/{user}/{repo}/tags")
+    if not r.ok:
+        return None
+
+    data = r.json()
+    tag_names = [t["name"] for t in data]
+    valid_tag_names = sorted(
+        [t for t in tag_names if semver.Version.is_valid(t.strip("v"))],
+        key=lambda t: semver.Version.parse(t.strip("v")),
+    )
+
+    if len(valid_tag_names) == 0:
+        return None
+
+    return valid_tag_names[-1]
+
+
+def _get_latest_pypi_version(name: str) -> str | None:
+    r = requests.get(f"https://pypi.org/pypi/{name}/json")
+    if not r.ok:
+        return None
+
+    data = r.json()
+    return data["version"]
+
+
+def _get_github_release_tag(url: str) -> str | None:
+    user, repo = url.split("/")[-2:]
+    r = requests.get(f"https://api.github.com/repos/{user}/{repo}/releases/latest")
+    if not r.ok:
+        return None
+
+    data = r.json()
+    return data["tag_name"]
+
+
+def get_version(config: SourceConfig) -> str:
+    """Get the latest version of the source.
+
+    Where the version comes from is defined in the config. If the version
+    should come from GitHub releases and no release is found, GitHub tags are
+    used as a fallback.
+
+    Args:
+        config (SourceConfig): configuration of the source
+
+    Returns:
+        str: semantic version string or "latest"
+    """
+    version = None
+
+    if config.version_by.version_type == VersionType.VERSION:
+        return config.version_by.value
+
+    if config.version_by.version_type == VersionType.PYPI:
+        version = _get_latest_pypi_version(config.version_by.value)
+        return version or "latest"
+
+    if config.version_by.version_type == VersionType.GH_RELEASE:
+        version = _get_github_release_tag(config.version_by.value)
+
+    if config.version_by.version_type == VersionType.GH_TAGS or version is None:
+        version = _get_latest_github_tag(config.version_by.value)
+
+    return version or "latest"
```

Comment:
I don't remember if we discussed to have `latest`?
Not super clear how to update it later